ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -227,123 +316,154 @@ prune maxSize st
-rewind slot st
-  | slot < anchor st = Nothing
-  | otherwise = Just $ PBftChainState {
-        anchor  = anchor st
-      , signers = keep
-      -- Typically we only drop a few blocks (and keep most), so updating
-      -- the counts based on what we drop will in most cases be faster than
-      -- recomputing it from the blocks we keep.
-      , counts  = repeatedly (decrementKey . pbftSignerGenesisKey)
-                             (toList discard)
-                             (counts st)
-      }
+-- In addition to preserving the invariant, we also have the guarantee that
+-- rolling back to a point (within @k@) and then reapplying the blocks that were
+-- rolled back results in the original state.
+rewind :: forall c. PBftCrypto c
+       => SecurityParam
+       -> WindowSize
+       -> WithOrigin SlotNo
+       -> PBftChainState c -> Maybe (PBftChainState c)
+rewind k n mSlot PBftChainState{..} =
+    case mSlot of
+      At slot ->
+        -- We scan from the right, since block to roll back to likely at end
+        case Seq.spanr (\(PBftSigner slot' _) -> slot' > slot) postAnchor of
+
+          -- We found the slot to roll back to post-anchor. Discard everything
+          -- after that slot.
+          (toDiscard, toKeep@(_ :|> x)) ->
+            if slot == pbftSignerSlotNo x
+              then Just $ go toDiscard toKeep
+              else notPreviouslyApplied
+
+          -- The slot was not found post-anchor. If the slot matches the last
+          -- slot pre-anchor, all is well, discarding everything post-anchor.
+          -- Otherwise, the rollback is too far.
+          (toDiscard, Empty) ->
+            case preAnchor of
+              _ :|> x
+                | slot == pbftSignerSlotNo x -> Just $ go toDiscard Empty
+                | slot <  pbftSignerSlotNo x -> rollbackToFar
+                | otherwise                  -> notPreviouslyApplied
+              _otherwise ->
+                notPreviouslyApplied
+
+      -- We can only roll back to origin if there are no signatures pre-anchor.
+      Origin ->
+        case preAnchor of
+          Empty      -> Just $ go postAnchor Empty
+          _otherwise -> rollbackToFar
   where
-    (keep, discard) = Seq.spanl (\(PBftSigner slot' _) -> At slot' <= slot) (signers st)
+    notPreviouslyApplied :: forall x. x
+    notPreviouslyApplied =
+      error $ "rewind: rollback to block not previously applied"
+
+    rollbackToFar :: Maybe x
+    rollbackToFar = Nothing
+
+    -- Construct new state, given the remaining post-anchor signatures
+    go :: StrictSeq (PBftSigner c)
+       -> StrictSeq (PBftSigner c)
+       -> PBftChainState c
+    go postAnchorDiscard postAnchorKeep = assertInvariant k n $ PBftChainState {
+          preAnchor  = preAnchor -- can't change by definition
+        , postAnchor = postAnchorKeep
+        , preWindow  = preWindow'
+        , inWindow   = inWindow'
+        , counts     = computeCounts inWindow' -- for simplicity, just recount
```

Comment:
Comment: We don't need to optimize this case as rollbacks in Byron are exceedingly rare.